docs/tooling-consolidation.md

```diff
@@ -0,0 +1,83 @@
+# AspNetCore Tooling Consolidation
+
+## Objectives
+
+We want to consolidate portions of dotnet/aspnetcore-tooling into dotnet/aspnetcore in our ongoing effort to reduce overall build time end to end for the .NET Core SDK. As such, we will not require all of aspnetcore-tooling to be merged into aspnetcore. For example, tooling for VSMac and VSCode will remain in aspnetcore-tooling. The main motivation is to achieve the overall goals of repo consolidation to reduce the number of repository required to build the SDK while keeping the build in aspnetcore as simple as possible without needing to support build/test for tooling scenarios such as testing on VSCode. There is also an added benefit of maintaining the current development and release workflow for aspnetcore-tooling which is more "agile" than aspnetcore (e.g. faster PR builds, release cycles that synchronize with VS releases).
```

Comment:
> For example, tooling for VSMac and VSCode will remain in aspnetcore-tooling.

Has this part of the plan had input from the Razor team?

 The ability to make compiler/language changes and see them running in the IDE is **essential** to the job. I realize this proposal implies the VS tooling with move with the compiler, so that's something. The more primary things like VSCode and VS4Mac become to our scenarios the more pain this causes. 

This means that doing a tooling release now means doing two builds across two repos and having package versions to keep in sync.

Reply:
We're definitely aiming to reduce end-to-end SDK build time in .NET 5.0. We're also trying to reduce the number of repos for consistency and ease of use / maintenance. I suggest dividing this document in two and clearly describing the goals of each part. For example, if we only care about runtime-to-SDK timing, why do anything after part 1? (What are the goals of part 2?)

Reply:
> It's definitely not ideal but at least for the interim it's a great stop gap and a preferred solution for us

You realize that this kills **dead** the ability to make language changes. There are some of those scheduled for .NET 5 for Blazor.

Reply:
I'm curious about the reasoning here. Why can't language changes be made after this if (as described below) aspnetcore-tooling gets an updated Razor SDK from aspnetcore?

Reply:
Put another way: With an additional goal to maintain the ability to test language changes immediately in the IDE, is it **required** to move everything from the aspnetcore-tooling repo's 'master' branch at the same time?

Reply:
@rynowak @NTaylorMullen Though there is a tradeoff between no cross-repo changes and faster PR validations, I don't see why all of aspnetcore-tooling **must** be kept together. Can one of you clarify on the requirements here?

Reply:
@rynowak is rightfully concerned about us doing Razor language changes and then losing the ability to immediately verify the tooling impact of those changes. I'm not saying that it will be impossible to validate the language changes, just much, much harder than today 😄. So that being said I 100% agree with Ryan that it sucks to have that restriction

; however, given the amount of non-language work we're currently pushing through and have planned

Basically i'm juggling the following priorities in my head:
1. Optimize build times for the dotnet org
2. Optimize the manual labor of "another repo"
3. Optimize stability for tooling specific changes (flakiness will heavily disrupt our ability to get out VS insertions)
4. Optimize speed for tooling specific changes. Re-building all of aspnetcore would not allow tooling to iterate quickly.
5. Optimize Razor language changes that also impact tooling.
6. Optimize Razor tooling only changes that do not require language changes.

As of right now we have a TON of 6 on the critical path as committed/currently being done. Hell that's a large reason why the aspnetcore/aspnetcore-tooling merge was delayed to begin with. And it goes without saying that 3 & 4 impact both language and non-language changes. Sooo all that said i'm willing to iteratively bring pieces of the aspnetcore-tooling world into AspNetCore to make the grander dotnet world better with the realization that there's pros and cons to it all. 

In an ideal world we'd "just" solve all of the above priorities. I haven't found a way to inexpensively do that.

Reply:
I wonder if there's a way to use submodules for local development scenarios that might make the experience more palatable. If building in CI, use packages, when building locally use ProjectReferences if they can be found.

Reply:
> Though there is a tradeoff between no cross-repo changes and faster PR validations

Languages changes for Blazor really can't be separated between *compiler change* and *tooling change* because it's just not architected that way. The bulk of the work for every language change is in tooling.

Reply:
@rynowak when you say "keeping everything building together", do you mean in CI or for local development?

@pranavkm I'm not sure how we could use submodules in one place but not the other unless we did the old aspnet/Universe thing and created another repo for Razor devs but never used it in pipelines.

----

Would we avoid these problems if we moved everything from aspnetcore-tooling into aspnetcore in one gulp? Is that possible and more palatable given the delay we're already dealing with?